### ISSUE TITLE
Request for feedback: Hydrodynamics-based pontoons/floats modelling

### ISSUE THREAD

**Issue by andgi** [Juanvvc/c172p-detailed#254]
My hydrodynamics-based (as opposed to contact point-based) modelling of floats is approaching an, I think, acceptable state. You can get it here
https://github.com/andgi/c172p-detailed
in the default hydrodynamics-wip branch. The branch will be rebased onto the upstream master branch frequently so if you add your own commits to it update it with git pull --rebase.

As far as I know the main thing missing now is completing the integration with the damage system and, of course, improving the hydrodynamic reactions - the latter is a long term effort and compared to a contact point based approach without any additional FDM parts (c.f. those in the Boeing314) I'd say it's already ahead.

Testing is needed, both to to discover problems in handling quality but also missing pieces with respect to the effects (spray, sparks, damage, etc, etc) present in the current contact based version.

Handling tips:
* On take-off a slight back-pressure on the stick is advisable, none or too much will get you into trouble. This effect has emerged from the modelling but it is unknown if it is in any way similar to the RL seaplane problem of porpoising which has a strong dependency on the pitch attitude.
* If the nose is anything more than a very tiny bit below the horizon when you touch down you will be very sorry - and wet.
* When the hydrodynamic reactions freak out, consider carefully what you did before and when that happened and if that would have been advisable to do in a RL seaplane. Would damage have occurred first?

**Comment by onox** [comment]
To test if you already have a clone of the Juanvvc/c172p-detailed repository, do this:

1. `git checkout -b andgi-master master`

2. `git pull https://github.com/andgi/c172p-detailed.git hydrodynamics-wip`

**Comment by onox** [comment]
Did some testing:

1. Didn't have as many difficulties as I had earlier. Keeping the nose above the horizon made the landings much less likely to freak out.

2. I saw a lot of white smoke, but couldn't see the grey smoke (what's the grey smoke for anyway?)

3. When I lowered the gear while trying to take off, neatly toppled the aircraft over after reaching > 40 knots:

![fgfs-screen-012](https://cloud.githubusercontent.com/assets/3361585/8022029/97c42fcc-0cb9-11e5-917e-6adfeb6e09df.png)

4. When I failed to do a takeoff 100 % correct, it was easy to damage one of the wings. Not sure if that's good or not.

Some screenshots:

![fgfs-screen-003](https://cloud.githubusercontent.com/assets/3361585/8022032/a1095828-0cb9-11e5-94f1-4bf9c85d6f6f.png)
![fgfs-screen-007](https://cloud.githubusercontent.com/assets/3361585/8022033/a10c139c-0cb9-11e5-9a20-3e7f2d14e288.png)
![fgfs-screen-010](https://cloud.githubusercontent.com/assets/3361585/8022031/a1097768-0cb9-11e5-87e3-af64bac9d29a.png)
![fgfs-screen-011](https://cloud.githubusercontent.com/assets/3361585/8022030/a109830c-0cb9-11e5-8297-72dab1b54e2f.png)


@wlbragg PS, I noticed a small problem with the instruments when they are in the shade or not:

![fgfs-screen-001](https://cloud.githubusercontent.com/assets/3361585/8022034/a8e4418e-0cb9-11e5-8e2c-51a344c6a9c0.png)

**Comment by onox** [comment]
@andgi Could you add (for future developers) a `README.md` file to `dev/fdm/gerris` explaining when and how you used `create_experiment.sh`, `run_experiment.sh`, and ` summarize_experiment_jsbsim.sh`. And how you generated `floats.gts.base`?

**Comment by tigert** [comment]
I can ask the guys who currently fly OH-CTL on floats about the procedures and practical handling. This sounds very interesting. CTL is not amphibious though (no wheels on floats)

**Comment by wkitty42** [comment]
FWIW it seems to me that amphibious and plain floats should be about the same other than the pocket for the rear wheels and if the front wheels are extended... maybe a few pounds more weight but not that much...

**Comment by tigert** [comment]
Yeah, I would assume the same, that operation on water is similar, except if you forget the wheels down on water :-P

I found a pretty short, but still useful guide on how the water operations work. Bear in mind I have no practical experience in these, but this document explains things well:

http://www.seaplanes.org/mambo/UserFiles/File/AC61-21A.pdf

Indeed it seems to be that full aft elevator is used on the beginning of takeoff to keep the nose up, and pressure is eased once the floats raise on the "step" and start planing.

**Comment by andgi** [comment]
Thanks guys, this helps identifying issues to work on (e.g. damage and crash detection - it is mostly the same as in master but I have change the gear/contact index where needed - and it needs more work to break things on water where no float contact points are used).

@onox I'll add a README. floats.gts.base is created from the float 3d model in floats.blend (same as in the main model but cleaned up to be a closed colume) by export to STL, then convert STL to gts with the gerris tool stl2gts. All table data in c172p-hydrodynamics come from the gerris experiments as created, run and summarized by the bash scripts in dev/fdm/gerris/.

@wkitty42 Currently, the hydrodynamics and contact points for amphibious and plain floats are the same. Their weight (and IIRC) in-air drag are different, which is inherited from the master branch.

**Comment by tigert** [comment]
For the record, I also have that shading issue on some instrument needles like on that screenshot, and this is new and recent, havent noticed it earlier.

**Comment by onox** [comment]
@andgi From which package do I get the "transform" binary?

**Comment by andgi** [comment]
@onox I built Gerris, http://gfs.sourceforge.net/wiki/index.php/Main_Page , from source last year and some times previously. I used sources from their darcs repository. I seem to have three source directories: gerris-stable  gfsview-stable  gts-stable

The transform command is probably part of the gts library. IIRC the command was called gts-transform when I had deb packages installed instead of building from source.

**Comment by onox** [comment]
@andgi It's indeed part of gts, but emerge only installs the man file here apparently, no binary. I guess I'll have to fix the .ebuild file then.

**Comment by wlbragg** [comment]
"I've noticed that the clearance between the forward (on ground) contact point and the runway is very small on the amphibious variant"
This still needs to be sorted out.

"I also have that shading issue on some instrument needles like on that screenshot, and this is new and recent, havent noticed it earlier."
This is probably the "internal shadow" inheritance that was applied to the instruments. It is definitely going to change the way they look if they are in shadow or sun. Unfortunately it can be a negative effect.

andgi, I have several issues with the new water FDM.
Things I really like are the floating effect and the front of the aircraft lifting slightly when power is applied The potential to nose dive (endo) is great!

Unfortunately, to me it appears really unstable right now. The main issues I noticed are...
Takes a long time to get enough speed to lift off, then I can't gain altitude and usually crash.
Towards the end of the roll out the aircraft starts to wobble, if not controlled perfectly it becomes uncontrollable very fast and then dances in a unrealistic way until it totally flips out.
I tried lowering the gear during roll out and it didn't appear to make any difference, maybe you don't have that input for that yet though.
Landing is also too touchy IMO, if you aren't just perfect it gets unrealistic in a hurry, (but I really like the endo when you do things wrong).
I hope it is possible to make this a little more stable because it has such great potential.
Am I the only one having these issues? Maybe I'm just really bad at piloting this aircraft with realistic FDM effects.

**Comment by wlbragg** [comment]
Another observation, just sitting dead on the water and turning left or right seems to influence the aircraft.
The floating and bobbing on the water is really cool, but the floats will sink completely below the water line periodically.

**Comment by wlbragg** [comment]
I've gained a little more control taking off, it definitely requires some touch!

**Comment by wlbragg** [comment]
The gear down on amphibious does make a difference but only if you put the nose down.

**Comment by wlbragg** [comment]
I know I am somewhat spamming this thread but after using this a little more, I really like the effort it takes to achieve a good takeoff and landing, it can be done, it just requires some effort. It feels so real when it's is done right. If we can just clean up the boundaries, it would be really nice.

**Comment by dany93** [comment]
@andgi wrote
>"I've noticed that the clearance between the forward (on ground) contact point and the runway is very small on the amphibious variant"
This still needs to be sorted out.

I'm looking at that (forward wheels on the ground).

@wlbragg wrote
>Takes a long time to get enough speed to lift off, then I can't gain altitude and usually crash.

After takeoff (not longer touching the water) do a bit of horizontal flight at this low height until you reach at least 80 kt, better 90 kts, before pulling the yoke to climb. After this, do not be below 80 - 90 kts.
The amphibious adds weight and drag, I have set them estimated or with the data that I found, hoping that some else would bring better. It obviously makes climbing more difficult. Like usually for every aircraft, the velocity (KIAS) is crucial. If it falls below 60 kts, and you pull the yoke to compensate the loss of climbing, you make things worse because of drag due to the angle attack. That is the"second regime", where you need more power to fly slower.
The propeller pitch angle may need some tuning, to pull more strongly at low velocity, less at high velocity. It is a compromise.
If to be seen, this is for later...

**Comment by tigert** [comment]
"climb prop" is no bad idea, old 172s are no speed kings on cruise anyway. Also ctl on floats has 180hp, 160hp on floats 172 is a very bad performing seaplane.

**Comment by dany93** [comment]
@tigert 
Yes, I remind your remark on the 180 hp for CTL.
@tigert wrote
 >172 is a very bad performing seaplane

This may be taken as a clue that the FDM is not so far from RL... Extra good point for JSBSim.

I take note for the future (recall me / us if needed) but this is some work to do ==> later in the calm, if we decide for it.
Also, this will be another FDM, I will not maintain two of them.

**Comment by dany93** [comment]
@wlbragg 
I have a model (currently in the state of working files, not clean) for the amphibious gears (height and suspension animation).
For you to check if my response is appropriate to your issue, what do you think is the best: I create a branch in my dany93 github or I do a Pull Request to juanvvc?

**Comment by wlbragg** [comment]
"I have a model ... for the amphibious gears"

Is there a reason you don't want to add it to your master, instead of branch or to juanvvc?

Any way you want to do it I can work with and check it out.

**Comment by onox** [comment]
Isn't it possible to control the hp of the engine with a property?

**Comment by dany93** [comment]
@wlbragg 
I didn't want to mess my master in case you would say " that was not my issue!".
*(cf. my hesitations with git....)*
Here is the link, dany93 rep, amphibious-gear branch https://github.com/dany93/c172p-detailed/tree/amphibious-gear

- Check the amphibious attitude and gears from ground,
- In the Fuel and payload dialog, you will find two loads for checking the suspension animation if you want.

front-load
rear-load

**Comment by dany93** [comment]
@onox 
>Isn't it possible to control the hp of the engine with a property?

Not that I know. But a file path would be good, it is a small file.
Moreover, the propeller is to be completely re-calculated, this is another file too.

**Comment by andgi** [comment]
@wlbragg I have not seen the issue with excessive bobbing up and down of the floats. Does it always happen? Or only sometimes? If the latter check the strength of the wind - the wings might add forces and moments that disturbs the equilibrium. I have also noticed that I have left a rudder yaw moment function from the MTB_20m in the hydrodynamics - I need to condition it on the water rudder being down.

@wlbragg  For the wobbling at the end of a landing run out - was that at slow speed? I'm not sure why that would happen. At higher speed wobbling (in pitch) is usually due to too high or too low pitch attitude which is qualitatively (but most likely not by degree) reasonable for a planing seaplane. If you wobble in roll while planing something introduced an unevenness in the load on the floats and then the deeper float will rebound and push down the other etc., so qualitatively that seems reasonable and would require corrective action by the pilot but for the magnitude and sensitivity, I don't know.

As an aside I can usually take off and land back successfully in external view, viewing the aircraft from the side, and as this view doesn't give very exact information on the roll and pitch attitude of the aircraft (compared to the cockpit view where you can judge it based on the nose v.s. horizon) I suspect that the hydrodynamics model might be more tolerant than reality or at least not excessively picky. It is harder and needs a more precise flare than landing on a runway, though.

I'll try to make some progress on the missing parts this evening.

**Comment by wlbragg** [comment]
andgi, amphibious gears are perfect, well done. Please send pull request to juanvvc with the changes. I think it would be easier for you to add than me and less prone to error.

**Comment by wlbragg** [comment]
The only thing I noticed that hopefully we can fix is when you hit g or shift+g the aircraft jumps to position then settles out appropriately. It may be the same issue I had with it during travel that I was able to work a fix for. We can look at it after we get it over to juanvvc. It is still superior to what we have now.

**Comment by dany93** [comment]
>
andgi, amphibious gears are perfect, well done. Please send pull request to juanvvc with the changes. I think it would be easier for you to add than me and less prone to error.

The files on my amphibious-gear branch https://github.com/dany93/c172p-detailed/tree/amphibious-gear were in a working state, ful of comments and with features (extra loads for tests) which must not be merged.
I can (intended to) clean them and do a pull request, but only to morrow. Please tell me how we are going to do, to avoid doing it twice.
If you do it prior to me, files involved (by memory):
physics.nas (Z values for [21[, [22], [23], [24] gears),
c172p.xml (spring coefficients),
Models/c172p.xml (wre lines 5068 - 5252, ```<!-- RETRACTABLE GEAR Ldamper, Rdamper -->```

**Comment by wlbragg** [comment]
I'll wait for you send pull request.

**Comment by wlbragg** [comment]
"I have not seen the issue with excessive bobbing up and down of the floats. Does it always happen? Or only sometimes?"
I thought it might have something to do with damage, but it turns out that is not the case. I can land on the water and just sit there dead in the water with no engine running and it starts to bob like a wave comes by, then it starts to sink and may stay submerged for as much as 5 or more seconds then it might pop up and stop doing it for awhile.
![floating](https://cloud.githubusercontent.com/assets/11069680/8046570/a5f1f8a6-0e01-11e5-8cb6-43434987fc8d.jpg)
![sunk](https://cloud.githubusercontent.com/assets/11069680/8046574/b22fe90c-0e01-11e5-9ed9-a12ef884ec75.jpg)

**Comment by wlbragg** [comment]
" For the wobbling at the end of a landing run out - was that at slow speed? "
I've now had it happen at both slow and high speed. I think it is related to "pitch attitude" and "roll attitude" as you suspect.

"I suspect that the hydrodynamics model might be more tolerant than reality or at least not excessively picky."
If this is true, remind me to avoid float planes because it doesn't take much of a mistake to loose it.
I was trying to take off a second ago and the run was taking too long so I tried to adjust pitch down then was going to slowly pitch up and I nose dived end over end. What a trip, I think it is so cool to have that possible outcome do to bad handling.

**Comment by wlbragg** [comment]
I have to say, I was really skeptical of adding this float FDM to master prior to freeze because I felt it was too unstable, but after using it some more, I am thinking it is suppose to be that way.
I am almost to the point of being convinced to merge this sooner rather than later so we have a few days to tune it with more eyes on it. What do you all think?

**Comment by onox** [comment]
@andgi Do you want to help adding [osgOcean](https://github.com/onox/osgocean) to FlightGear?

**Comment by wkitty42** [comment]
if we're shooting for the 12th, then best to get this in now and get it tested... otherwise, leave it for 3.8... that gives a 5 day buffer before the freeze...

**Comment by andgi** [comment]
@wlbragg I'll have to look closer at the bobbing issue. The floats should not sink like that and the indicated airspeed is low so there isn't much wind from the front, at least. When this happens can you check the wind properties to see if there is a strong wind from the side or from aft? The FDM beta and airspeed properties should give this information (these are not part of the on-screen display).

@wlbragg I have not changed anything functional with the landing gears of the amphibious variant, the only change is that the their indices have changed as I removed two of the pontoon contact points as they were repeated for both variants with floats while the current float model is the same for both so there is no reason to have duplicates. OTOH I put 3 contact points per float: fore, step and aft. Their Z-location is managed in Systems/bushkit.xml rather than in Nasal/physics as was the case before.

I need to put in overstress detection for the water reactions and prepare a merge request branch. Or perhaps the other way around?

**Comment by andgi** [comment]
@onox I might have some time to look at osgOcean during my vacation in July, but there are so many things on my todo list. A key feature of my current patch of water approach is that I can replicate the water surface both in the shader and in the FDM so that the physics can react to what you see. The 3d part of this is not included in the c172p, but the FDM part is, so if you set a wave amplitude in /fdm/jsbsim/hydro/environment/ you will see the aircraft react to invisible waves. The 3d part is still unready so it cannot be put in FGData for the primary aircraft yet.

**Comment by wlbragg** [comment]
"I removed two of the pontoon contact points as they were repeated for both variants with floats while the current float model is the same for both so there is no reason to have duplicates."
I think I duplicated them to use one set to control floating on land (if I remember correctly). But the FDM changes would eliminate that necessity anyway.
I need to remember to add the extra contact points to spray, wake, spark, splash and anywhere else I use the contact points.

**Comment by tigert** [comment]
i think this is reasonable and nice when you land properly, and the swells can be very nice when you add a bit of waves in the props.

Issue for me still is the inability of the amphibious model to move. Main wheels are halfway in asphalt and it doesnt roll at all.

**Comment by dany93** [comment]
![c172p-amphib-wheels-2015-06-10 11 16 58](https://cloud.githubusercontent.com/assets/11994061/8079102/b67d943e-0f62-11e5-8857-e993d064c9f7.png)
Version PR and merged yesterday on juanvvc:master (wlbragg's amphibious, usual FDM contacts).
It rolls and takes off with no problem.
Even the previous version rolled.

**Comment by andgi** [comment]
@dany93 Yes, the new gears look fine, but the hydrodynamics-wip branch is not up to date with that yet. Additionally, something broke physics.nas last time I rebased onto master so with the current hydrodynamics-wip branch anything that depends on physics.nas is broken. I will try to fix it ASAP but might not have time until tomorrow.

@tigert Try with some head wind and with the yoke back - then you might lift the offending structure contact point off the ground and get rolling. Alternatively, open System/bushkit.xml and raise the z-position of contact point [13] and [14]. You should be able to find the components that set the contact point height towards the end of the file.

**Comment by andgi** [comment]
The hydrodynamics-wip branch should be ok again and up to date with master now.

**Comment by onox** [comment]
@andgi Is there anything that needs to be done before you can/want to create a pull request? In case you have modified some contact points, you might want to update `Systems/flight-recorder/components/damage.xml` or `Systems/flight-recorder/components/gear-amphibious.xml`.

**Comment by andgi** [comment]
@onox Thanks, I'll look at those. Otherwise I think damage detection is the major missing part. I'll try to prepare a pull request this evening.

**Comment by onox** [comment]
> I also need to look at the water sounds.

You mean #264?

**Comment by andgi** [comment]
Yes, I saw the sound files being added, but didn't know that they were not yet used in the c172-sound.xml file. The conditions for the sounds will be different depending on whether floats on water are implemented with contact points or the hydrodynamics system. I suppose /fdm/jsbsim/hydro/active-norm and /fdm/jsbsim/hydro/v-kt will be the two most useful properties for configuring the water sounds.

**Comment by onox** [comment]
Good to know the conditions are different. #264 can be fixed after your branch has been merged.

**Comment by tigert** [comment]
Yeah, now it rolls, and it is fun to fly. And since I have flown on fixed gear aircraft only, of course I took off with the amphibian from EFHN, and headed south to the sea, and got a quick swim by forgetting the gear down on landing and tumbling upside down on touchdown ;-)

This is very fun stuff, thank you very much for your efforts :-) 

One weird thing I noticed, is that when the gear is lowered on water, the aircraft seemed to move in a circle, only turning left for some reason. I need to test this a bit more (and the 180HP version, the extra 30 horsepower are really useful on the load version, it's really really easy to get to the "too slow" flight regime and just sink into the trees if you really are not careful - 160HP is quite dangerously underpowered for the floats.

**Comment by andgi** [comment]
With gear (BOGEY) contact points down JSBSim has built in water drag which I don't know if I can control and it behaves somewhat strange (try taking off from water with the gear down). In the worst case, if the built in gear drag is not behaving well enough we can move the gears up when on water and substitute our own drag forces and moments, but most likely not before the freeze, though.

**Comment by wlbragg** [comment]
"In the worst case, if the built in gear drag is not behaving well enough we can move the gears up when on water and substitute our own drag forces and moments, but most likely not before the freeze, though."
I haven't looked to see how much (if any) of the code that does just what your talking about you have removed in physics.nas, but the condition as to "over" land VS water is there. Then again that may not be where you need to adjust "everything" from, especially the FDM portion.

**Comment by andgi** [comment]
@wlbragg I moved/added that logic for the float contact points (now only for land use) to bushkit.xml as it might be better if these contact points are moved immediately when they sense (or not) solid ground. With Nasal there might be a delay of a number of FDM iterations. I could do the same for the amphibious landing gears - it is not difficult. OTOH, extended landing gears should have some effect in the water and the default JSBSim behaviour provides that but in (AFAIK) a non-tuneable way.

**Comment by onox** [comment]
> I could do the same for the amphibious landing gears

That would be nice.

**Comment by wlbragg** [comment]
"extended landing gears should have some effect in the water"
Yes, and right now I really like some of their effects (endo) :)

Off Topic
I am used to working directly from a local clone, In this case I have branched your repo in github, do your master repo changes automatically appear in my branch or do I have to update my "online" branch somehow?

**Comment by andgi** [comment]
I have filed a pull request now. It is basically the state from yesterday but I have squashed most of the internal development, disabled the debug display, and updated affected flight-recorder configuration files.

**Comment by tigert** [comment]
Some more footage provided by the "CTL guys" :-)

https://www.youtube.com/watch?v=FS_8lzcxrhc

They'll be interested in testing the behavior once I get around figuring out controls for the sim they can use - mouse yoke is not ideal ..
